### DISCUSSION TITLE
Refactoring to use new SDK under the hood

### DISCUSSION THREAD

**Discussion by Dominic-Marcelino** [Intevel/nuxt-directus#173] [General]
The nuxt-module hasn't been relying on the directus SDK in order to do things the nuxt-way. With the new SDK coming soon (currently in beta) I'm wondering if it wouldn't make sense to re-evaluate this.

The new SDK has better types support and relies on fetch instead of axios.

With building on top of the SDK developer would have benetifs like:

- More consistend DX between Nuxt and non-Nuxt Projects communicating with Directus 
- Feature parity (the module is still missing a buch of functions like file-upload, getFileByID)
- Tree-shakable features (only bundle what you need)

Looking forward to your feedback

**Comment by casualmatt** [comment]
I can't speak for @Intevel, but the idea, in general, is that if we migrate from $fetch (unJS ofetch) we do it for useFetch or useAsyncData to better integrate with Nuxt Multiple Environment support, which will not grant the support for fetch, and could introduce edge cases.

**Reply by casualmatt** [reply]
And you can already expose Directus SDK as a plugin yourself if you want the whole SDK experience, but as for Strapi the module is made to fit the Nuxt ecosystem.

**Reply by Dominic-Marcelino** [reply]
> but the idea, in general, is that if we migrate from $fetch (unJS ofetch) we do it for useFetch or useAsyncData

`useFetch` and `useAsyncData` both use a caching system, which is not wished in all use-cases. I assume that's why this module also uses the pure `$fetch`. 

If it's only about fetch, it would be possible to override the fetch object. Right now it's only possible for the global fetch, but I'm pretty sure the team is open for discussion in order to pass a  custom fetch instance / wrapper.
https://discord.com/channels/725371605378924594/1121094628037361768/1121114471423823953

**Reply by casualmatt** [reply]
I need to check but I don't think that the 2 composables introduce any form of caching, they ar just optimized for various rendering mode, so that request are not repeates on hydration.

In any case we didn't proceed in migrating to useFetch because of braking changes that will introduce.

**Reply by sandros94** [reply]
Yes, it is mainly for hydration consistency and basic-state management of the fetched data. But more importantly `useFetch` is basically `useAsync($fetch)`.

If fact this is how I use it when I need consistency in hydration or reduce at maximum the api calls:

```ts
const { data: VARIABLE-NAME } = await useAsyncData('KEY-NAME', () =>
  getItems({
    [...]
  })
)
```

This way in a subsequent page (or component) that needs the same data I could just do:

```ts
const VARIABLE-NAME = useNuxtData('KEY-NAME')
```

Or even use it like a *cache* to [preview something for that data](https://nuxt.com/docs/api/composables/use-nuxt-data#show-stale-data-while-fetching-in-the-background) that needs frequents updates even on client side.

**Reply by casualmatt** [reply]
That is what we suggest currently, but we already implement `$fetch` instead of `fetch `to improve support for different rendering modes.
@Intevel and I are working to understand how the new SDK, could be implemented in Nuxt with Nuxt `UseFetch `or `UseAsyncData `to support all rendering modes better and not have duplicate requests. 

But this could take some time; for now, we still support the current module, and a new update should come in a few days.

**Comment by Intevel** [comment]
@Dominic-Marcelino Im looking forward to maybe rewrite this module using the new SDK & the powerfull features provided by Nuxt. 

We will track this process in this discussion.

**Comment by Intevel** [comment]
Im currently trying something out with the new SDK, you can check https://github.com/intevel/nuxt-directus/tree/dev branch. And may contribute if you want, we check if the SDK suits. @Dominic-Marcelino

**Reply by sandros94** [reply]
I would like to help, but I feel a little bit lost since this is only the third module I work on. To start with I'm reading the `@directus/sdk` documentation.

How are you planning to do this? Looking at the code I imagine that you are exposing the full sdk via `useDirectus` composable and then build a set of other composables with the same scopes as the currently available ones? (eg. I see `useDirectusAuth` and `useDirectusCookie`)

**Reply by Intevel** [reply]
The idea is to wrap some of the functions of the SDK (as well as Auth) into own composables to continue using Nuxt features like `useCookie`.

**Reply by sandros94** [reply]
> The idea is to wrap some of the functions of the SDK (as well as Auth) into own composables to continue using Nuxt features like `useCookie`.

Speaking of `useCookie`: since there have been recent changes ([nuxt 3.6.0 release](https://github.com/nuxt/nuxt/releases/tag/v3.6.0) that included some [state management](https://github.com/nuxt/nuxt/pull/20970) for it), alongside with other (potentially useful) functionalities, would you consider to bump the nuxt version of the module and the related minimum requirement? Or are you aiming to make the module fully retro compatible with all nuxt 3 versions?

EDIT: actually I don't fully know if using a higher version of nuxt to create the module would result into the need of bumping the minimum requirement of nuxt version where the module gets used

**Reply by sandros94** [reply]
@Intevel idk about you but I'm loving this sdk (never used the old one).

It is so simple to use and I'll be happy to try my best and rewrite the current `nuxt-directus` module with it.
I'm opening a draft PR so you could check the progress (already made the auth work and optimized the cookies a bit), I'll much appreciate your opinion.

**Comment by Dominic-Marcelino** [comment]
I saw that some initial work has been done. Has there been any outcome weather the approach is promising and will be continued?

**Reply by sandros94** [reply]
Ehy @Dominic-Marcelino, I took some day off from everything, just as a summer break. My plan is to get back on creating a couple more composables to have a somewhat trustable comparison between the two.

I think that only then @Intevel could say something about it and decide.

IMHO there are a number of already built in feature in the sdk that I much enjoy and are still missing from the original module (like logged user self-update method #74). Not to mention that for each missing composable, the rewrite with the new sdk exposes it directly to nuxt. So developers that are using the rewritten module, who might require a specific missing feature, could simply code it by themselves and maybe contribute them with already existing code/testing.

In the end not only I see some small benefits here and there, but even a reduction in dev time (the need of recreating logics already available in the sdk). But again, I'm sure @Intevel might have a broader view on the topic.

**Reply by casualmatt** [reply]
I can give an idea of what I and @Intevel have discussed, but I can't speak for him.

- The main concern is duplicated requests using the SDK if you have ssr enabled.
- The SDK don't support SSG or other rendering option. 
- The SDK uses fetch and not unJS oFetch. 
- Talking with Directus to understand what we could do to make the SDK more flexible or how we could use it to better fit Nuxt ecosystem. 


This, instead, is my take: 
If you want to use the SDK, import it as a nuxt plugin. A module should support/integrate with nuxt on a deeper level, with the ability to turn SSR, ISR, or noJS on.

**Reply by sandros94** [reply]
Mainly replying just to discuss the concepts of a potential rewrite (while understanding a bit better how this module currently works) and sharing more of my opinions on the topic:

> - The main concern is duplicated requests using the SDK if you have ssr enabled.
> - [...]
> - The SDK uses fetch and not unJS oFetch.

Isn't this the case with the current state too? In #148, #149 and #153 wasn't this about discussing the use of [`useAsyncData`](https://nuxt.com/docs/api/composables/use-async-data) to prevent duplicated requests? (from the nuxt docs about `useAsyncData`: *handler: an asynchronous function that returns a value*. Regardeless of a fetch or oFetch)

> - The SDK don't support SSG or other rendering option.

On this one I'm a bit lost because I have zero to low experience other than SSR, but how could this be a problem? To my understanding the SDK doesn't have any restriction on where it can be used (lately I've noticed an astro SSG POC using it, but currently couldn't find the repo, so I cannot double check if it indeed works or not).

> - Talking with Directus to understand what we could do to make the SDK more flexible or how we could use it to better fit Nuxt ecosystem.

Surely I cannot, nor want to, speak for them, but the last time this discussion came out with @rijkvanzanten my understanding was that they (Directus team) prefer to make the SDK as generic as possible, without focusing on per-framework optimization. Operation that should be done by the community or the framework interested in doing so. As per the Nuxt & Directus collaboration as of now (documentation, videos and webinar made together) looks to be more like "writing down correct guides and use cases".

### Personal notes
In general, while I do agree with your take on:
> [...] A module should support/integrate with nuxt on a deeper level, [...]
In this situation I'm against it. Let me explain my point of view:

This module should be the main goto for devs that would like to build projects with Nuxt and Directus right? And as such it should reflect as much as possible both docs and tooling, while providing those extra functionalities that would work as the [keystone](https://en.wikipedia.org/wiki/Keystone_(architecture)), taking the advantages of both sides.

Why? 
- Easier maintainability of the module.
- final Developer Experience

The first would be achieved by the community, more or less like it is done now, but with the overview and occasional help from the devs of **both** Nuxt and Directus teams.
The second focusing on the maintainability of the projects where this module could be used. To be more precise on this with a couple of use cases:
1. Lets take the current state of the module in project `X` (not ex-twitter, just an hypothetical project). The devs that work on `X` are used to `nuxt-directus` but need a new functionality that is not present in the module but is already documented in the official directus docs. This means that the main tools that they chosen to work with (Nuxt and Directus) could achieve their goal with their respective tools, but the one supposed to do so doesn't (yet, hypothetically). The solution in that moment? Install the SDK and import it as a plugin. But then they face two problems:
   - they have to maintain the code for both nuxt-directus and SDK in their project `X` (personally something I do only for short terms or POC)
   - they probably wont contribute to `nuxt-directus` module with their solution, because this would require *reinventing the wheel* by recreating the required logic inside `nuxt-directus` that is already available in the SDK. Making the last point even harder to do in the long run.
2. On the other hand, lets take in consideration project `Y`, built with `directus-sdk` on top of a non-nuxt framework. They decide to move to Nuxt for whatever built-in reasons. While for sure they will keep using the `directus-sdk` they will be probably be hesitant from using `nuxt-directus` because that would involve similar maintenance stories described in the last points. Again, potentially preventing useful future contributions to this module.

---

I really hope my reply didn't sound like I'm forcing my opinion. I only would like to understand more your point of view and explain better mine. ❤️

**Reply by casualmatt** [reply]
More context:
We know that we will rewrite the code of the module with the new useFetch or useAsyncData from next that ware not available at the time.


I will give a full reply later but I'm also open for a call on discord.

**Reply by rijkvanzanten** [reply]
> Surely I cannot, nor want to, speak for them, but the last time this discussion came out with @rijkvanzanten my understanding was that they (Directus team) prefer to make the SDK as generic as possible, without focusing on per-framework optimization.

This is correct! If there's any changes to the SDK that would make life easier to frameworks in general, then do absolutely let me know. If it's an addition/tweak that only works with Nuxt exclusively, I'd rather have that added in on the Nuxt module side of things 💪🏻

**Reply by sandros94** [reply]
> We know that we will rewrite the code of the module with the new useFetch or useAsyncData from next that ware not available at the time.

I remember something along those lines being discussed, in fact it is something I would like to replicate it in the dev-only draft-PR I'm writing. Making sense of the "keystone concept", but yet making sure that there aren't disadvantages of doing so.

> I will give a full reply later

Would like to read it, once you have time to write it.

> but I'm also open for a call on discord.

I tried to add you using your gh user, but got the wrong guy. It's probably easier if you add me instead (same username and picture).

**Reply by casualmatt** [reply]
We are evaluating what to do: 
- Rewrite with useFetch the current module 
or 
- Rewrite with the SDK
 
The module was written by @Intevel when Nuxt was in beta, not even in RC, and many features were unavailable. 
Now nuxt offers useFetch, and because it's a native feature of nuxt, we are sure that it will work in every environment that nuxt now and in the future will support (those are always expanding; let's think about lagon added when the platform was closed beta) and we don't know if those platforms will support native fetch or will introduce differences in how the fetch API behaves, but we know that useFetch will support them. 
So rewriting everything using that would ensure no issues on the environment side, and we could optimize the module to use as many nuxt features as possible with no compatibility issues. 
The second big advantage is you should almost always use useAsyncData with the SDK if you use any rendering engine other than SPA. This is because it will double the amount of API calls without it, or in some cases, could not even work without the composable around it. 
Support for multiple environments and rendering technologies makes Nuxt amazing, with the hybrid possibilities of the nitro server, using ssg for some pages and ssr or spa for others. 

The SDK offers a lot of good advantages, such as features already developed by Directus, ts out-of-the-box, and it will always be updated with directus, but then we still have to build around it a compatibility layer for those alternative renderings like SSG that need useAsyncData(you should always use it not to have a duplicate request), so there is still a lot of work to do anyway to make it fit nuxt ecosystem. 

I know that is not a complete list of every topic, but it's a better idea of what we are testing publicly and privately on my side and what I have discussed with Connor.

**Reply by Dominic-Marcelino** [reply]
I think my motivation for using the sdk under the hood are mostly the same as @Sandros94 already brought up. Working with the current module (at least for me) it feels like it tries to replicate most of the extisting features with a slightly different syntax. If the benefit is there, that also would be fine. But this brings up the main problem: feature parity. When rewriting to much of the existing functionality there will never be a support for all new features which makes it hard to keep with the module. Also this requires to write documentations, which already exist for the official sdk.
For example there's currently no out of the box support to upload images or setup 2fa. Also the module could greatly benefit from the new Typing-Feature, especially when the auto type-generation will be released. 

All in one I think that there should definitely be a go-to module (like this), but with feature parity and a similar DX like the official module.

Based from what I read the mail problem is still based on the `fetch` vs `oFetch`. As mentioned before I could imagine that a solution to provide a custom fetch instance could be discussed with the directus team (@rijkvanzanten) :-)

**Reply by sandros94** [reply]
> Also this requires to write documentations, which already exist for the official sdk.

The module must have its documentation. Regardless of the approach we take `nuxt-directus` must integrate the vast majority of the Nuxt related features that @casualmatt described (in particular rendering modes). But yes, with a potential rewrite using the sdk it could mean not only having the ability to use it as a backup (auto imported it), but even access new features as soon the sdk dep is updated, even before being wrapped into nuxt-optimized composables.

> Based from what I read the mail problem is still based on the `fetch` vs `oFetch`.

This is the one topic I feel least informed on. I don't even know what `oFetch` provides compared to `fetch`. I must understand it first.

**Reply by rijkvanzanten** [reply]
> Based from what I read the mail problem is still based on the fetch vs oFetch. As mentioned before I could imagine that a solution to provide a custom fetch instance could be discussed with the directus team (@rijkvanzanten) :-)

Great example of the type of stuff that's a great tweak to the SDK that makes it better for any framework 😄

**Reply by casualmatt** [reply]
@Sandros94, here is some extra info on ofetch:
ofetch is the fetch library from unJS, offering the same dev experience on any platform that Nitro officially supports. 
So, it is important to ensure that at least the module will always work on any platform officially supported by Nuxt/Nitro.

Different it's to use useFetch, which combines ofetch and useasyncdata and ensures that all rendering and nuxt API will work, like SSR,SSG, refresh all data, ...

**Reply by Intevel** [reply]
@Dominic-Marcelino Thats right, and that would be exactly the reasons to use the SDK, many features are not yet implemented in this module, but in the SDK. Types would always have to be maintained here, and ....

My thought in the current is to use the SDK and implement auth and other core features myself or in combination with the SDK and Nuxt features.

**Reply by Dominic-Marcelino** [reply]
FYI: there's an open PR from the team to configure fetch: https://github.com/directus/directus/pull/19606

**Reply by rijkvanzanten** [reply]
Closed PR now 😄 (It'll ship today!)

**Reply by sandros94** [reply]
> Closed PR now 😄 (It'll ship today!)

Oh man, this opens up so many things to do!

**Reply by Intevel** [reply]
Looks nice, thanks.

**Comment by sandros94** [comment]
Just as an update, I opened up a new discussion to track the current state of the rewrite at #215. If any of you have considerations on the topic or want to follow the progress I encourage you to go and comment or subscribe to that discussion! 😁

**Reply by Dominic-Marcelino** [reply]
Thanks for your great work! 🎉

**Comment by casualmatt** [comment]
--> https://github.com/Intevel/nuxt-directus/discussions/271

nuxt-directus-next is EOL. 

We will start to work on a v6 based on Nuxt First Approach
